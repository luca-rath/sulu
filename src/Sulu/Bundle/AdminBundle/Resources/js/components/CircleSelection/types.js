// @flow
export interface Normalizer {
    normalize(data: SelectionData): SelectionData,
}

export type SelectionData = {
    left: number,
    radius: number,
    top: number,
};
