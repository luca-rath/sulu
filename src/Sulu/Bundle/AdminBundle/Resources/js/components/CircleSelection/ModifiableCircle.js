// @flow
import type {ElementRef} from 'react';
import React from 'react';
import {action, observable} from 'mobx';
import {observer} from 'mobx-react';
import classNames from 'classnames';
import type {SelectionData} from './types';
import modifiableCircleStyles from './modifiableCircle.scss';

type Props = {
    disabled: boolean,
    filled: boolean,
    label: string | typeof undefined,
    left: number,
    onChange?: (value: SelectionData) => void,
    onDoubleClick?: () => void,
    radius: number,
    resizable: boolean,
    top: number,
};

@observer
class ModifiableCircle extends React.Component<Props> {
    @observable moveMode = false;
    @observable resizeMode = false;
    @observable resizeAngle = 0;

    circleRef: ?ElementRef<'div'>;

    componentDidMount() {
        window.addEventListener('mouseup', this.handleMouseUp);
        window.addEventListener('mousemove', this.handleMouseMove);
    }

    componentWillUnmount() {
        window.removeEventListener('mouseup', this.handleMouseUp);
        window.removeEventListener('mousemove', this.handleMouseMove);
    }

    setCircleRef = (ref: ?ElementRef<'div'>) => {
        this.circleRef = ref;
    };

    @action handleMoveMouseDown = (event: MouseEvent) => {
        event.stopPropagation();
        this.moveMode = true;
    };

    @action handleResizeMouseDown = (event: MouseEvent) => {
        event.stopPropagation();
        this.resizeMode = true;
    };

    @action handleMouseUp = () => {
        this.moveMode = false;
        this.resizeMode = false;
    };

    @action handleMouseMove = (event: MouseEvent) => {
        const {onChange, radius, left, top} = this.props;

        if (this.moveMode) {
            const {movementX, movementY} = event;

            if (!onChange) {
                return;
            }

            onChange({
                radius,
                left: left + movementX,
                top: top + movementY,
            });
        }

        if (this.resizeMode) {
            if (!this.circleRef) {
                return;
            }

            const rect = this.circleRef.getBoundingClientRect();

            const circleX = rect.left + rect.width / 2;
            const circleY = rect.top + rect.height / 2;
            const {clientX: mouseX, clientY: mouseY} = event;

            const deltaX = mouseX - circleX;
            const deltaY = mouseY - circleY;

            this.resizeAngle = (Math.atan2(deltaY, deltaX) * 180) / Math.PI;

            if (!onChange) {
                return;
            }

            const newRadius = Math.sqrt(deltaX ** 2 + deltaY ** 2);

            onChange({
                left,
                top,
                radius: newRadius,
            });
        }
    };

    handleDoubleClick = this.props.onDoubleClick;

    render() {
        const {disabled, filled, resizable, label, radius, left, top} = this.props;
        const width = !resizable && radius === 0 ? 30 : radius * 2;
        const labelSize = radius === 0 ? 14 : Math.sqrt(radius) * 5;

        const circleClass = classNames(
            modifiableCircleStyles.circle,
            {
                [modifiableCircleStyles.disabled]: disabled,
                [modifiableCircleStyles.filled]: filled,
            }
        );

        return (
            <div
                className={circleClass}
                onDoubleClick={!disabled && this.handleDoubleClick || undefined}
                onMouseDown={!disabled && this.handleMoveMouseDown || undefined}
                ref={this.setCircleRef}
                role="button"
                style={{
                    left: left + 'px',
                    top: top + 'px',
                    width: width + 'px',
                    height: width + 'px',
                }}
            >
                {label &&
                    <div
                        className={modifiableCircleStyles.label}
                        style={{fontSize: `${labelSize}px`}}
                    >
                        {label}
                    </div>
                }
                {resizable && !disabled &&
                    <div
                        className={modifiableCircleStyles.resizeHandle}
                        onMouseDown={this.handleResizeMouseDown}
                        role="slider"
                        style={{
                            transformOrigin: `calc(50% + ${radius * -1}px) 50%`,
                            transform: `translate(calc(-50% + ${radius}px), -50%) rotate(${this.resizeAngle}deg)`,
                        }}
                    />
                }
            </div>
        );
    }
}

export default ModifiableCircle;
