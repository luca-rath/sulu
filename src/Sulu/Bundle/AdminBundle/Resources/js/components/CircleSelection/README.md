This is a general purpose component to select a circle area on all kind of content. The properties allow to specify a
minimum radius as well as a maximum radius, which the selection box cannot undercut. A double click on the selection box
centers it and maximizes its size, if it's resizable.

```javascript
const initialSelection = {radius: 100, left: 50, top: 50};

const [selection, setSelection] = React.useState(initialSelection);

<div>
    <CircleSelection
        onChange={setSelection}
        value={selection}
    >
        <div>
            <p style={{padding: '20px'}}>Lorem ipsum dolor sit amet, consetetur sadipscing elitr, sed diam nonumy eirmod tempor invidunt ut labore et dolore magna aliquyam erat, sed diam voluptua. At vero eos et accusam et justo duo dolores et ea rebum. Stet clita kasd gubergren, no sea takimata sanctus est Lorem ipsum dolor sit amet. Lorem ipsum dolor sit amet, consetetur sadipscing elitr, sed diam nonumy eirmod tempor invidunt ut labore et dolore magna aliquyam erat, sed diam voluptua. At vero eos et accusam et justo duo dolores et ea rebum. Stet clita kasd gubergren, no sea takimata sanctus est Lorem ipsum dolor sit amet.</p>
        </div>
    </CircleSelection>

    {selection &&
        <p>
            Radius: {selection.radius}<br />
            Left: {selection.left}<br />
            Top: {selection.top}
        </p>
    }

    <button onClick={() => setSelection(initialSelection)}>Reset to initial value</button>
</div>
```

Content placed inside the component is not allowed to change it's size after rendering.  Otherwise, the behaviour will
be undefined. For example when rendering images, they need to be preloaded before rendering the selection.

```javascript
// preload image
let image = new Image();
image.src = 'https://unsplash.it/800/500';

const [imageLoaded, setImageLoaded] = React.useState(image.complete);
const [selection, setSelection] = React.useState({radius: 200, top: 100, left: 100});

image.onload = () => setImageLoaded(true);

imageLoaded
    ? <div>
        <CircleSelection
            minRadius={100}
            maxRadius={200}
            onChange={setSelection}
            value={selection}
            label="1"
        >
            <img src="https://unsplash.it/800/500"/>
        </CircleSelection>

        <p>
            Radius: {selection.radius}, 
            Top: {selection.top}, 
            Left: {selection.left}
        </p>
    </div>
    : null
```

If the values should be in percent, set `usePercentageValues={true}`. Be aware, that the `radius` is relative to the `containerWidth` then.

```javascript
// preload image
let image = new Image();
image.src = 'https://unsplash.it/800/500';

const [imageLoaded, setImageLoaded] = React.useState(image.complete);
const [selection, setSelection] = React.useState({radius: 0.1, top: 0.5, left: 0.5});

image.onload = () => setImageLoaded(true);

imageLoaded
    ? <div>
        <CircleSelection
            label="1"
            minRadius={0.06}
            onChange={setSelection}
            usePercentageValues={true}
            value={selection}
        >
            <img src="https://unsplash.it/800/500"/>
        </CircleSelection>

        <p>
            Radius: {selection.radius},
            Top: {selection.top}, 
            Left: {selection.left}
        </p>
    </div>
    : null
```

There is also the possibility to render a PointSelection by setting `resizable={false}` and omitting the radius.

```javascript
// preload image
let image = new Image();
image.src = 'https://unsplash.it/800/500';

const [imageLoaded, setImageLoaded] = React.useState(image.complete);
const [selection, setSelection] = React.useState({top: 100, left: 100});

image.onload = () => setImageLoaded(true);

imageLoaded
    ? <div>
        <CircleSelection
            filled={true}
            label="55"
            resizable={false}
            onChange={setSelection}
            value={selection}
        >
            <img src="https://unsplash.it/800/500"/>
        </CircleSelection>

        <p>
            Top: {selection.top}, 
            Left: {selection.left}
        </p>
    </div>
    : null
```

To disable a CircleSelection, set `disabled={true}`.

```javascript
// preload image
let image = new Image();
image.src = 'https://unsplash.it/800/500';

const [imageLoaded, setImageLoaded] = React.useState(image.complete);
const [selection, setSelection] = React.useState({radius: 100, top: 200, left: 200});

image.onload = () => setImageLoaded(true);

imageLoaded
    ? <div>
        <CircleSelection
            disabled={true}
            onChange={setSelection}
            value={selection}
        >
            <img src="https://unsplash.it/800/500"/>
        </CircleSelection>

        <p>
            Radius: {selection.radius}, 
            Top: {selection.top}, 
            Left: {selection.left}
        </p>
    </div>
    : null
```

There is also the possibility, to render multiple CircleSelections for one single container. But you should use the
`withContainerSize` hoc to automatically pass `containerWidth` and `containerHeight` to the `CircleSelectionRenderer`
components. For the sake of simplicity it's hardcoded in this example.

```javascript
// preload image
let image = new Image();
image.src = 'https://unsplash.it/800/500';

const [imageLoaded, setImageLoaded] = React.useState(image.complete);
const [firstSelection, setFirstSelection] = React.useState({radius: 200, top: 100, left: 100});
const [secondSelection, setSecondSelection] = React.useState({radius: 100, top: 150, left: 500});
const [active, setActive] = React.useState(1);

image.onload = () => setImageLoaded(true);

imageLoaded
    ? <div>
        <div style={{ width: '800px', height: '500px', position: 'relative', overflow: 'hidden', display: 'inline-flex' }}>
            <img src="https://unsplash.it/800/500" style={{ userSelect: 'none', pointerEvents: 'none' }} alt="Unsplash image" />
            <CircleSelection.Renderer
                disabled={active !== 1}
                onChange={setFirstSelection}
                value={firstSelection}
                label="1"
                containerWidth={800}
                containerHeight={500}
            />
            <CircleSelection.Renderer
                disabled={active !== 2}
                onChange={setSecondSelection}
                value={secondSelection}
                label="2"
                containerWidth={800}
                containerHeight={500}
            />
        </div>
        <div>
            <p>Active: {active}</p>
            <button onClick={() => {setActive(1)}}>#1</button>
            <button onClick={() => {setActive(2)}}>#2</button>
        </div>
    </div>
    : null
```
