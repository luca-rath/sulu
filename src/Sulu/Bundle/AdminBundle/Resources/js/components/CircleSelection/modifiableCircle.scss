@import '../../containers/Application/colors.scss';

$backgroundColor: rgba($black, .6);
$borderColor: $white;
$borderSize: 2px;
$backgroundColorFilled: $white;
$backgroundColorFilledDisabled: rgba($backgroundColorFilled, .6);
$boxShadowColorFilled: rgba($black, .6);
$labelColor: $white;
$labelColorDisabled: rgba($labelColor, .5);
$labelColorFilled: $black;
$labelColorFilledDisabled: rgba($labelColorFilled, .6);
$resizeHandleBoxShadowColor: rgba($black, .3);

.circle {
    position: absolute;
    background: $backgroundColor;
    border-radius: 100%;
    transform: translate(-50%, -50%);
    will-change: width, height, left, top;
    cursor: move;
    border: $borderSize solid $borderColor;

    &.disabled {
        cursor: default;
        border: none;
        pointer-events: none;
    }

    &.filled {
        background: $backgroundColorFilled;
        box-shadow: 0 0 2px 1px $boxShadowColorFilled;

        &.disabled {
            background: $backgroundColorFilledDisabled;
        }
    }
}

.label {
    position: absolute;
    left: 50%;
    top: 50%;
    transform: translate(-50%, -50%);
    font-weight: bold;
    color: $labelColor;
    user-select: none;

    .circle.disabled & {
        color: $labelColorDisabled;
    }

    .circle.filled & {
        color: $labelColorFilled;
    }

    .circle.filled.disabled & {
        color: $labelColorFilledDisabled;
    }
}

.resize-handle {
    position: absolute;
    width: 20px;
    height: 20px;
    border-radius: 10px;
    left: 50%;
    top: 50%;
    transform: translate(-50%, -50%);
    background: $borderColor;
    cursor: nwse-resize;
    box-shadow: 0 0 20px 0 $resizeHandleBoxShadowColor;
    will-change: transform, transform-origin;
}
