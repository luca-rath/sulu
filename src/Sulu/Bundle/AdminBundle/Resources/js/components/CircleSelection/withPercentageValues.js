// @flow
import React from 'react';
import type {ComponentType} from 'react';
import {observer} from 'mobx-react';
import {computed} from 'mobx';
import {buildHocDisplayName} from '../../utils/react';
import type {SelectionData} from './types';

type Props = {
    containerHeight: number,
    containerWidth: number,
    maxRadius: number | typeof undefined,
    minRadius: number | typeof undefined,
    onChange: (value: ?SelectionData) => void,
    usePercentageValues: boolean,
    value: SelectionData | typeof undefined,
};

export default function withPercentageValues(Component: ComponentType<*>) {
    @observer
    class WithPercentageValuesComponent extends React.Component<Props> {
        wrappedComponent = Component;

        static defaultProps = {
            maxRadius: undefined,
            minRadius: undefined,
            usePercentageValues: false,
        };

        handleChange = (value: ?SelectionData) => {
            const {containerHeight, containerWidth, onChange} = this.props;

            if (!value) {
                onChange(value);

                return;
            }

            onChange({
                ...value,
                left: value.left / containerWidth,
                top: value.top / containerHeight,
                radius: value.radius / containerWidth,
            });
        };

        @computed get transformedMaxRadius() {
            const {containerWidth, maxRadius} = this.props;

            if (!maxRadius) {
                return maxRadius;
            }

            return maxRadius * containerWidth;
        }

        @computed get transformedMinRadius() {
            const {containerWidth, minRadius} = this.props;

            if (!minRadius) {
                return minRadius;
            }

            return minRadius * containerWidth;
        }

        @computed get transformedValue() {
            const {containerHeight, containerWidth, value} = this.props;

            if (!value) {
                return value;
            }

            return {
                ...value,
                left: value.left * containerWidth,
                top: value.top * containerHeight,
                radius: value.radius * containerWidth,
            };
        }

        render() {
            const {usePercentageValues} = this.props;

            if (!usePercentageValues) {
                return (
                    <Component {...this.props} />
                );
            }

            const props = {
                ...this.props,
                maxRadius: this.transformedMaxRadius,
                minRadius: this.transformedMinRadius,
                onChange: this.handleChange,
                value: this.transformedValue,
            };

            return (
                <Component {...props} />
            );
        }
    }

    WithPercentageValuesComponent.displayName = buildHocDisplayName('withPercentageValues', Component);

    return WithPercentageValuesComponent;
}
