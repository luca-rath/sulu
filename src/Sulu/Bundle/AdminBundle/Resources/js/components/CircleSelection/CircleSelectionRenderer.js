// @flow
import React from 'react';
import type {Normalizer, SelectionData} from './types';
import ModifiableCircle from './ModifiableCircle';
import PositionNormalizer from './normalizers/PositionNormalizer';
import RoundingNormalizer from './normalizers/RoundingNormalizer';
import SizeNormalizer from './normalizers/SizeNormalizer';

type Props = {
    containerHeight: number,
    containerWidth: number,
    disabled: boolean,
    filled: boolean,
    label?: string,
    maxRadius?: number,
    minRadius?: number,
    onChange: (value: ?SelectionData) => void,
    resizable: boolean,
    round: boolean,
    usePercentageValues: boolean,
    value: SelectionData | typeof undefined,
};

export default class CircleSelectionRenderer extends React.Component<Props> {
    static defaultProps = {
        disabled: false,
        filled: false,
        resizable: true,
        round: true,
        usePercentageValues: false,
    };

    handleChange = (value: ?SelectionData) => {
        const {onChange, containerWidth, containerHeight, usePercentageValues} = this.props;

        if (!usePercentageValues || !value) {
            onChange(value);

            return;
        }

        const {left, top, radius = 0} = value;

        onChange({
            left: left / containerWidth,
            top: top / containerHeight,
            radius: radius / containerWidth,
        });
    };

    get value() {
        const {value, containerWidth, containerHeight, usePercentageValues} = this.props;

        if (!value) {
            return this.getMaximumSelection();
        }

        if (!usePercentageValues) {
            return value;
        }

        const {left, top, radius = 0} = value;

        return {
            left: left * containerWidth,
            top: top * containerHeight,
            radius: radius * containerWidth,
        };
    }

    componentDidMount() {
        this.setInitialValue();
    }

    componentDidUpdate() {
        this.setInitialValue();
    }

    setInitialValue = () => {
        const {value} = this.props;

        if (!value) {
            this.handleChange(this.value);
        }
    };

    static createNormalizers(props: Props): Array<Normalizer> {
        const {containerWidth, containerHeight, maxRadius, minRadius, usePercentageValues, round, resizable} = props;

        if (!containerWidth || !containerHeight) {
            return [];
        }

        const normalizers = [
            new PositionNormalizer(
                containerWidth,
                containerHeight
            ),
        ];

        if (resizable) {
            const calculatedMaxRadius = usePercentageValues && maxRadius ? maxRadius * containerWidth : maxRadius;
            const calculatedMinRadius = usePercentageValues && minRadius ? minRadius * containerWidth : minRadius;

            normalizers.push(
                new SizeNormalizer(
                    containerWidth,
                    containerHeight,
                    calculatedMaxRadius,
                    calculatedMinRadius
                )
            );
        }

        if (round && !usePercentageValues) {
            normalizers.push(new RoundingNormalizer());
        }

        return normalizers;
    }

    get normalizers() {
        return CircleSelectionRenderer.createNormalizers(this.props);
    }

    normalize(selection: SelectionData): SelectionData {
        return this.normalizers.reduce((data, normalizer) => normalizer.normalize(data), selection);
    }

    getMaximumSelection = (): SelectionData => {
        const {containerWidth, containerHeight, resizable} = this.props;

        const radius = (containerWidth && containerHeight && resizable)
            ? Math.min(containerWidth, containerHeight) / 2
            : 0;

        return this.normalize(
            this.centerSelection({
                left: 0,
                top: 0,
                radius,
            })
        );
    };

    centerSelection(selection: SelectionData): SelectionData {
        const {containerWidth, containerHeight} = this.props;

        const halfWidth = containerWidth / 2;
        const halfHeight = containerHeight / 2;

        return {
            ...selection,
            left: halfWidth,
            top: halfHeight,
        };
    }

    handleCircleDoubleClick = () => {
        const {resizable} = this.props;

        if (resizable) {
            this.handleChange(this.getMaximumSelection());

            return;
        }

        this.handleChange(this.normalize(this.centerSelection(this.value)));
    };

    handleCircleChange = (value: SelectionData) => {
        this.handleChange(this.normalize(value));
    };

    render() {
        const {disabled, resizable, label, filled} = this.props;
        const {left, top, radius = 0} = this.value;

        return (
            <ModifiableCircle
                disabled={disabled}
                filled={filled}
                label={label}
                left={left}
                onChange={this.handleCircleChange}
                onDoubleClick={this.handleCircleDoubleClick}
                radius={radius}
                resizable={resizable}
                top={top}
            />
        );
    }
}
