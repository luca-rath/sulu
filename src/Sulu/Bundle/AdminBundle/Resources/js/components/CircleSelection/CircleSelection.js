// @flow
import type {Node} from 'react';
import React from 'react';
import withContainerSize from '../withContainerSize';
import CircleSelectionRenderer from './CircleSelectionRenderer';
import type {SelectionData} from './types';
import circleSelectionStyles from './circleSelection.scss';

type Props = {
    children?: Node,
    disabled: boolean,
    filled: boolean,
    label?: string,
    maxRadius?: number,
    minRadius?: number,
    onChange: (value: ?SelectionData) => void,
    resizable: boolean,
    round: boolean,
    value: SelectionData | typeof undefined,
};

class CircleSelection extends React.Component<Props & {
    containerHeight: number,
    containerWidth: number,
}> {
    static defaultProps = {
        disabled: false,
        filled: false,
        resizable: true,
        round: true,
    };

    render() {
        const {
            children,
            containerHeight,
            containerWidth,
            disabled,
            filled,
            label,
            maxRadius,
            minRadius,
            onChange,
            resizable,
            round,
            value,
        } = this.props;

        return (
            <div className={circleSelectionStyles.selection}>
                {children}
                <CircleSelectionRenderer
                    containerHeight={containerHeight}
                    containerWidth={containerWidth}
                    disabled={disabled}
                    filled={filled}
                    label={label}
                    maxRadius={maxRadius}
                    minRadius={minRadius}
                    onChange={onChange}
                    resizable={resizable}
                    round={round}
                    value={value}
                />
            </div>
        );
    }
}

export {
    CircleSelection,
};

export default class CircleSelectionComponent extends React.Component<Props> {
    static defaultProps = CircleSelection.defaultProps;

    static Renderer = CircleSelectionRenderer;

    render() {
        const Component = withContainerSize(CircleSelection, circleSelectionStyles.container);

        return <Component {...this.props} />;
    }
}
