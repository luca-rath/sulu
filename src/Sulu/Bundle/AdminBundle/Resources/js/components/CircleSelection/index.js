// @flow
import CircleSelection from './CircleSelection';
import type {SelectionData} from './types';

export default CircleSelection;
export type {SelectionData};
