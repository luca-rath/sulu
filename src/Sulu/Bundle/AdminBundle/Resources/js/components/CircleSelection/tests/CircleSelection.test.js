// @flow
import {mount} from 'enzyme';
import React from 'react';
import {CircleSelection} from '../CircleSelection';
import CircleSelectionRenderer from '../CircleSelectionRenderer';

jest.mock('../../../utils/Translator', () => ({
    translate: jest.fn((key) => key),
}));

jest.mock('../../withContainerSize/withContainerSize');
jest.mock('../../../utils/DOM/afterElementsRendered');

test('The component should render', () => {
    const view = mount(
        <CircleSelection
            containerHeight={1000}
            containerWidth={2000}
            onChange={jest.fn()}
            value={{radius: 1, top: 3, left: 4}}
        >
            <p>Lorem ipsum</p>
        </CircleSelection>
    );

    expect(view.render()).toMatchSnapshot();
});

test('The component should center and maximize the selection if no value is given', () => {
    const changeSpy = jest.fn();

    mount(
        <CircleSelection containerHeight={1000} containerWidth={2000} onChange={changeSpy} value={undefined}>
            <p>Lorem ipsum</p>
        </CircleSelection>
    );

    expect(changeSpy).toBeCalledWith({left: 1000, radius: 500, top: 500});
});

test('The component should reset the value if modifiable circle is doubleclicked', () => {
    const changeSpy = jest.fn();

    const view = mount(
        <CircleSelection
            containerHeight={1000}
            containerWidth={2000}
            minHeight={100}
            minWidth={100}
            onChange={changeSpy}
            value={{radius: 100, top: 30, left: 40}}
        >
            <p>Lorem ipsum</p>
        </CircleSelection>
    );

    view.find('ModifiableCircle').prop('onDoubleClick')();

    expect(changeSpy).toBeCalledWith({left: 1000, radius: 500, top: 500});
});

test('The component should center and maximize the selection when a minRadius and maxRadius is given', () => {
    const changeSpy = jest.fn();

    mount(
        <CircleSelection
            containerHeight={1000}
            containerWidth={2000}
            maxRadius={200}
            minRadius={50}
            onChange={changeSpy}
            value={undefined}
        >
            <p>Lorem ipsum</p>
        </CircleSelection>
    );

    expect(changeSpy).toBeCalledWith({left: 1000, radius: 200, top: 500});
});

test('The component should not allow the selection to move over the borders', () => {
    const changeSpy = jest.fn();

    const view = mount(
        <CircleSelection
            containerHeight={1000}
            containerWidth={2000}
            onChange={changeSpy}
            value={{left: 0, top: 0, radius: 2000}}
        >
            <p>Lorem ipsum</p>
        </CircleSelection>
    );

    view.find(CircleSelectionRenderer).first().children().first().instance().handleCircleChange(
        {radius: 0, left: -10, top: -20}
    );
    expect(changeSpy).toBeCalledWith({radius: 0, top: 0, left: 0});
});

test('The component should not allow the selection to be bigger than the container', () => {
    const changeSpy = jest.fn();

    const view = mount(
        <CircleSelection
            containerHeight={1000}
            containerWidth={2000}
            onChange={changeSpy}
            value={{left: 0, top: 0, radius: 2000}}
        >
            <p>Lorem ipsum</p>
        </CircleSelection>
    );

    view.find(CircleSelectionRenderer).first().children().first().instance().handleCircleChange(
        {radius: 5000, left: 0, top: 0}
    );
    expect(changeSpy).toBeCalledWith({radius: 2236, top: 0, left: 0});
});

test('The component should not round if told by the properties', () => {
    const changeSpy = jest.fn();

    const view = mount(
        <CircleSelection
            containerHeight={1000}
            containerWidth={2000}
            minRadius={5}
            onChange={changeSpy}
            round={false}
            value={undefined}
        >
            <p>Lorem ipsum</p>
        </CircleSelection>
    );

    expect(view.find('ModifiableCircle').prop('top')).toBeGreaterThan(166);
    expect(view.find('ModifiableCircle').prop('left')).toBeGreaterThan(166);
    expect(view.find('ModifiableCircle').prop('radius')).toBeGreaterThan(166);
});

test('The component should work with percentage values if told by the properties', () => {
    const changeSpy = jest.fn();

    mount(
        <CircleSelection
            containerHeight={1000}
            containerWidth={2000}
            onChange={changeSpy}
            usePercentageValues={true}
            value={undefined}
        >
            <p>Lorem ipsum</p>
        </CircleSelection>
    );

    expect(changeSpy).toBeCalledWith({top: 0.5, left: 0.5, radius: 0.25});
});
