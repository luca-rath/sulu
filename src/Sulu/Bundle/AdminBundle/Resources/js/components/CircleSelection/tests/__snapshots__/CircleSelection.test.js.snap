// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The component should center and maximize the selection when a minRadius and maxRadius is given 1`] = `
<div
  class="selection"
>
  <p>
    Lorem ipsum
  </p>
  <div
    class="circle"
    role="button"
    style="left: 1000px; top: 500px; width: 400px; height: 400px;"
  >
    <div
      class="resizeHandle"
      role="slider"
      style="transform-origin: calc(50% + -200px) 50%; transform: translate(calc(-50% + 200px), -50%) rotate(0deg);"
    />
  </div>
</div>
`;

exports[`The component should render with children 1`] = `
<div
  class="selection"
>
  <p>
    Lorem ipsum
  </p>
  <div
    class="circle"
    role="button"
    style="left:1000px;top:500px;width:1000px;height:1000px"
  >
    <div
      class="resizeHandle"
      role="slider"
      style="transform-origin:calc(50% + -500px) 50%;transform:translate(calc(-50% + 500px), -50%) rotate(0deg)"
    />
  </div>
</div>
`;

exports[`The component should render with value as selection 1`] = `
<div
  class="selection"
>
  <p>
    Lorem ipsum
  </p>
  <div
    class="circle"
    role="button"
    style="left: 4px; top: 3px; width: 2px; height: 2px;"
  >
    <div
      class="resizeHandle"
      role="slider"
      style="transform-origin: calc(50% + -1px) 50%; transform: translate(calc(-50% + 1px), -50%) rotate(0deg);"
    />
  </div>
</div>
`;
